src/sentry/db/models/outboxes.py

```diff
@@ -0,0 +1,141 @@
+        raise NotImplementedError
+
+
+class ProcessUpdatesWithRegionOutboxes(ProducesRegionOutboxOnUpdate):
+    __category__: OutboxCategory
+    __outbox_type__: Type[RegionOutboxBase] | None = None
+
+    def payload_for_update(self) -> Mapping[str, Any] | None:
+        """
+        A custom json payload to be included in outboxes generated via creation, update, or deletion.
+        Note that outboxes are COALESCED!  This means that when multiple updates are processed at once,
+        only the latest value is used and all others ignored.  That means that not every payload generated is
+        guaranteed to be processed.
+        """
+        return None
+
+    def outbox_for_update(self, shard_identifier: int | None = None) -> RegionOutboxBase:
+        """
+        Returns outboxes that result from this model's creation, update, or deletion.
+        Subclasses generally should override outbox_region_names or payload_for_update to customize
+        this behavior.
+        """
+        return self.__category__.as_region_outbox(
+            model=self,
+            payload=self.payload_for_update(),
+            shard_identifier=shard_identifier,
+            outbox=self.__outbox_type__,
+        )
+
+    @classmethod
+    def handle_async_deletion(
+        cls, identifier: int, shard_identifier: int, payload: Mapping[str, Any] | None
+    ) -> None:
+        """
+        Called one or more times as an outbox receiver processes the class update category and the
+        given identifier is not found in the database.  This method can be used to invoke service
+        methods to destroy cross silo resources not already handled by HybridCloudForeignKey (those
+        are handled via maybe_process_tombstone).  Note that this method can be called many times --
+        for instance, failure to process the outbox will result in a retry.  Thus, all operations
+        in this method must be entirely idempotent and safe to async / stale states that can occur.
+        """
+        pass
+
+    def handle_async_replication(self, shard_identifier: int) -> None:
+        """
+        Called one or more times as an outbox receiver processes the class update category and
+        the given identifier is found in the database.  This method can be used to invoke service
+        methods to replicate, enable, or alter cross silo resources based on updates to this model.
+        Note that because outboxes are COALESCED this means that not every unique update will be processed.
+        Also keep in mind that any errors or failures will force a retry of processing, so be certain all
+        operations are idempotent!
+        """
+        pass
+
+
+class ProducesControlOutboxesOnUpdate:
+    __default_flush__ = True
+
+    def outboxes_for_update(self, shard_identifier: int | None = None) -> List[ControlOutboxBase]:
+        raise NotImplementedError
+
+
+class ProcessUpdatesWithControlOutboxes(ProducesControlOutboxesOnUpdate):
```

Comment:
I kind of want `ProcessUpdatesWith*Outboxes` to share a common abstract superclass, seeing that they share four method names. If I'm eyeballing this correctly:

* `payload_for_update` is identical for both;
* `outboxes_for_update` has the same signature but a different body; and
* the signatures for `outboxes_for_update`, `handle_async_deletion`, and `handle_async_replication` differ in whether they take a `region_name` argument.

If there were value in having a common interface, we could solve the `region_name` problem by extracting those methods into a helper object that's created after receiving a region name if one is needed. That is, the call sites might look like

```python
model.get_outbox_processor().handle_async_replication(123)  # if control model
model.get_outbox_processor_for_region("na").handle_async_replication(123)  # if region model
```

But I recognize that's more of a Java-ish design pattern and might not be worth the extra lines of code in Python.

Reply:
Some important notes though:

1.  `outboxes_for_update` only applies to the ProcessUpdatesWithControlOutboxes, not the region one.  Only Control outboxes have region names and only Control models produce multiple outboxes per update.  Note the subtle signature and name difference.
2.  I do see your point.  I think part of the trick, however, is that I wanted the `override` for `handle_async_replication` to exist on the ORM because it's "dead simple" and in line with other ORM connected interfaces.  Given that the interface of the mixin itself informs an implementor (and their IDE) of the presence or lack of a region name, it is very simple to approach it with separate method signatures without the intermediate object.  Thinking of it this way: there is no common implementation of `handle_async_replication` that would work in both region and control modes (even the tombstone service either does or does not require the region parameter, for example).